File: com.trollworks.gcs/src/com/trollworks/gcs/ui/scale/Scale.java
/*
 * Copyright ©1998-2020 by Richard A. Wilkes. All rights reserved.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, version 2.0. If a copy of the MPL was not distributed with
 * this file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This Source Code Form is "Incompatible With Secondary Licenses", as
 * defined by the Mozilla Public License, version 2.0.
 */

package com.trollworks.gcs.ui.scale;

import com.trollworks.gcs.ui.UIUtilities;

import java.awt.Component;
import java.awt.Font;
import java.awt.Insets;

/** Provides convenience for scaling. */
public class Scale {
    private static Scale  OVERRIDE;
    private        double mScale;

    public static void setOverride(Scale scale) {
        OVERRIDE = scale;
    }

    /**
     * @param comp The component to determine the scale for.
     * @return The scale.
     */
    public static Scale get(Component comp) {
        if (OVERRIDE != null) {
            return OVERRIDE;
        }
        ScaleRoot root  = UIUtilities.getSelfOrAncestorOfType(comp, ScaleRoot.class);
        Scale     scale = null;
        if (root != null) {
            scale = root.getScale();
        }
        return scale != null ? scale : Scales.ACTUAL_SIZE.getScale();
    }

    /**
     * Creates a new scale.
     *
     * @param scale The scale to use, where 1.0 is 100%, 2.0 is 200%, etc.
     */
    public Scale(double scale) {
        mScale = scale;
    }

    /** @return The scale, where 1.0 is 100%, 2.0 is 200%, etc. */
    public double getScale() {
        return mScale;
    }

    /**
     * @param font The font to scale.
     * @return The scaled font.
     */
    public Font scale(Font font) {
        if (mScale == 1) {
            return font;
        }
        return font.deriveFont((float) (font.getSize() * mScale));
    }

    /**
     * @param insets The insets to scale.
     * @return The scaled insets.
     */
    public Insets scale(Insets insets) {
        if (mScale == 1) {
            return insets;
        }
        return new Insets((int) (insets.top * mScale), (int) (insets.left * mScale), (int) (insets.bottom * mScale), (int) (insets.right * mScale));
    }

    /**
     * @param values The values to scale.
     * @return The scaled values.
     */
    public double[] scale(double[] values) {
        if (mScale == 1) {
            return values;
        }
        int      length = values.length;
        double[] scaled = new double[length];
        for (int i = 0; i < length; i++) {
            scaled[i] = values[i] * mScale;
        }
        return scaled;
    }

    /**
     * @param value The value to scale.
     * @return The scaled value.
     */
    public int scale(int value) {
        return (int) (value * mScale);
    }

    /**
     * @param value The value to scale.
     * @return The scaled value.
     */
    public float scale(float value) {
        return (float) (value * mScale);
    }

    /**
     * @param value The value to scale.
     * @return The scaled value.
     */
    public double scale(double value) {
        return value * mScale;
    }
}

File: ux/scale.go
// Copyright (c) 1998-2026 by Richard A. Wilkes. All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, version 2.0. If a copy of the MPL was not distributed with
// this file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// This Source Code Form is "Incompatible With Secondary Licenses", as
// defined by the Mozilla Public License, version 2.0.

package ux

import (
	"github.com/richardwilkes/gcs/v5/model/gurps"
	"github.com/richardwilkes/toolbox/v2/geom"
	"github.com/richardwilkes/toolbox/v2/i18n"
	"github.com/richardwilkes/toolbox/v2/xreflect"
	"github.com/richardwilkes/unison"
	"github.com/richardwilkes/unison/enums/mod"
)

// ScaleDelta is the delta used when adjusting the view scale incrementally.
const ScaleDelta = 10

// NewScaleField creates a new scale field and hooks it into the target.
func NewScaleField(minValue, maxValue int, defValue, get func() int, set func(int), afterApply func(), attemptCenter, adjustForDisplayPPI bool, scroller *unison.ScrollPanel) *PercentageField {
	applyFunc := func() {
		scale := float32(get()) / 100
		if adjustForDisplayPPI {
			scale *= float32(gurps.GlobalSettings().General.MonitorPPI()) / 72
		}
		scalePt := geom.NewPoint(scale, scale)
		if header := scroller.ColumnHeader(); !xreflect.IsNil(header) {
			header.AsPanel().SetScale(scalePt)
		}
		scroller.Content().AsPanel().SetScale(scalePt)
	}
	scaleTitle := i18n.Text("Scale")
	overrideCenter := false
	var override geom.Point
	var scaleField *PercentageField
	scaleField = NewPercentageField(nil, "", scaleTitle, get,
		func(scale int) {
			if scaleField != nil && !scaleField.Enabled() {
				SetFieldValue(scaleField.Field, scaleField.Format(get()))
				return
			}
			var view, content *unison.Panel
			var viewRect geom.Rect
			var center geom.Point
			var dX, dY float32
			if attemptCenter {
				view = scroller.ContentView()
				viewRect = view.ContentRect(false)
				content = scroller.Content().AsPanel()
				if overrideCenter {
					center = override
				} else {
					center = viewRect.Center()
				}
				dX = center.X - viewRect.X
				dY = center.Y - viewRect.Y
				center = content.PointFromRoot(view.PointToRoot(center))
			}

			set(scale)
			applyFunc()
			scroller.MarkForLayoutRecursively()
			scroller.ValidateLayout()

			if content != nil {
				fScale := float32(scale) / 100
				viewRect.Width /= fScale
				viewRect.Height /= fScale
				viewRect.X = center.X - dX/fScale
				viewRect.Y = center.Y - dY/fScale
				content.ScrollRectIntoView(viewRect)
			}

			if afterApply != nil {
				afterApply()
			}
		}, minValue, maxValue, false, false)
	scaleField.SetMarksModified(false)
	scaleField.Tooltip = newWrappedTooltip(scaleTitle)
	scroller.ContentView().MouseWheelCallback = func(where, delta geom.Point, mods mod.Modifiers) bool {
		if !mods.OptionDown() || !scaleField.Enabled() {
			return false
		}
		current := get()
		scale := current + int(delta.Y*ScaleDelta)
		if scale < minValue {
			scale = minValue
		} else if scale > maxValue {
			scale = maxValue
		}
		if current != scale {
			overrideCenter = true
			override = where
			SetFieldValue(scaleField.Field, scaleField.Format(scale))
			overrideCenter = false
		}
		return true
	}
	installFunc := func() {
		scroller.ParentChangedCallback = nil
		installViewScaleHandlers(unison.Ancestor[unison.Dockable](scroller), minValue, maxValue, get, defValue,
			func(scale int) {
				if get() != scale {
					SetFieldValue(scaleField.Field, scaleField.Format(scale))
				}
			})
		applyFunc()
		if afterApply != nil {
			afterApply()
		}
	}
	if dockable := unison.Ancestor[unison.Dockable](scroller); xreflect.IsNil(dockable) {
		scroller.ParentChangedCallback = installFunc
	} else {
		installFunc()
	}
	return scaleField
}

func installViewScaleHandlers(paneler unison.Paneler, minValue, maxValue int, current, def func() int, adjuster func(scale int)) {
	p := paneler.AsPanel()
	installViewScaleHandler(p, ScaleDefaultItemID, minValue, maxValue, current, def, adjuster)
	installViewDeltaScaleHandler(p, ScaleUpItemID, ScaleDelta, minValue, maxValue, current, adjuster)
	installViewDeltaScaleHandler(p, ScaleDownItemID, -ScaleDelta, minValue, maxValue, current, adjuster)
	installViewScaleHandler(p, Scale25ItemID, minValue, maxValue, current, func() int { return 25 }, adjuster)
	installViewScaleHandler(p, Scale50ItemID, minValue, maxValue, current, func() int { return 50 }, adjuster)
	installViewScaleHandler(p, Scale75ItemID, minValue, maxValue, current, func() int { return 75 }, adjuster)
	installViewScaleHandler(p, Scale100ItemID, minValue, maxValue, current, func() int { return 100 }, adjuster)
	installViewScaleHandler(p, Scale200ItemID, minValue, maxValue, current, func() int { return 200 }, adjuster)
	installViewScaleHandler(p, Scale300ItemID, minValue, maxValue, current, func() int { return 300 }, adjuster)
	installViewScaleHandler(p, Scale400ItemID, minValue, maxValue, current, func() int { return 400 }, adjuster)
	installViewScaleHandler(p, Scale500ItemID, minValue, maxValue, current, func() int { return 500 }, adjuster)
	installViewScaleHandler(p, Scale600ItemID, minValue, maxValue, current, func() int { return 600 }, adjuster)
}

func installViewScaleHandler(p *unison.Panel, itemID, minValue, maxValue int, current, desired func() int, adjuster func(int)) {
	scale := desired()
	if minValue <= scale && maxValue >= scale {
		p.InstallCmdHandlers(itemID,
			func(_ any) bool {
				d := desired()
				return minValue <= d && maxValue >= d && current() != d
			},
			func(_ any) { adjuster(desired()) })
	}
}

func installViewDeltaScaleHandler(p *unison.Panel, itemID, delta, minValue, maxValue int, current func() int, adjuster func(int)) {
	calc := func() int {
		c := current()
		adjusted := c + delta
		if adjusted < minValue {
			adjusted = minValue
		} else if adjusted > maxValue {
			adjusted = maxValue
		}
		return adjusted
	}
	p.InstallCmdHandlers(itemID, func(_ any) bool { return calc() != current() }, func(_ any) { adjuster(calc()) })
}
